src/mame/drivers/firebeat.cpp

```diff
@@ -865,368 +1127,381 @@ WRITE_LINE_MEMBER(firebeat_state::spu_ata_dmarq)
-	map(0x800000, 0xffffff).rw(FUNC(firebeat_state::firebeat_waveram_r), FUNC(firebeat_state::firebeat_waveram_w));
+void firebeat_bm3_state::firebeat_bm3_map(address_map &map)
+{
+	firebeat_spu_map(map);
+
+	map(0x7d000330, 0x7d00033f).nopw(); // ?
+	map(0x7d000340, 0x7d00035f).r(FUNC(firebeat_bm3_state::sensor_r));
+	map(0x70001fc0, 0x70001fdf).rw(FUNC(firebeat_bm3_state::fdd_unk_r), FUNC(firebeat_bm3_state::fdd_unk_w));
+	map(0x70008000, 0x7000807f).r(FUNC(firebeat_bm3_state::spectrum_analyzer_r));
+}
+
+uint32_t firebeat_bm3_state::spectrum_analyzer_r(offs_t offset)
+{
+	// Visible in the sound test menu and most likely the spectral analyzer game skin
+	//
+	// Notes about where this could be coming from...
+	// - It's not the ST-224: Only sends audio in and out, with a MIDI in
+	// - It's not the RF5C400: There are no unimplemented registers or anything of that sort that could give this info
+	// - The memory address mapping is the same as Keyboardmania's wheel, which plugs into a connector on extend board
+	//   but there's nothing actually plugged into that spot on a beatmania III configuration, so it's not external
+	// - Any place where the audio is directed somewhere (amps, etc) does not have a way to get back to the PCBs
+	//   from what I can tell based on looking at the schematics in the beatmania III manual
+	// - I think it's probably calculated somewhere within one of the main boards (main/extend/SPU) but couldn't find any
+	//   potentially interesting chips at a glance of PCB pics
+	// - The manual does not seem to make mention of this feature *at all* much less troubleshooting it, so no leads there
+
+	// 6 notch spectrum analyzer
+	// No idea what frequency range each notch corresponds but it does not affect core gameplay in any way.
+	// Notch 1: 0x0c
+	// Notch 2: 0x0a
+	// Notch 3: 0x08
+	// Notch 4: 0x06
+	// Notch 5: 0x04
+	// Notch 6: 0x02
+
+	// TODO: Fill in logic (reuse vgm_visualizer in some way?)
+	// auto ch = offset & 0xf0; // 0 = Left, 1 = Right
+	auto data = 0;
+
+	return data;
+}
+
+uint16_t firebeat_bm3_state::sensor_r(offs_t offset)
+{
+	switch (offset) {
+		case 0: return ioport("IO1")->read() | 0x0100;
+		case 1: return ioport("IO2")->read() | 0x0100;
+		case 2: return ioport("IO3")->read() | 0x0100;
+		case 3: return ioport("IO4")->read() | 0x0100;
+		case 5: return (ioport("TURNTABLE_P1")->read() >> 8) | 0x0100;
+		case 6: return (ioport("TURNTABLE_P1")->read() & 0xff) | 0x0100;
+		case 7: return (ioport("TURNTABLE_P2")->read() >> 8) | 0x0100;
+		case 8: return (ioport("TURNTABLE_P2")->read() & 0xff) | 0x0100;
+		case 9: return ioport("EFFECT1")->read() | 0x0100;
+		case 10: return ioport("EFFECT2")->read() | 0x0100;
+		case 11: return ioport("EFFECT3")->read() | 0x0100;
+		case 12: return ioport("EFFECT4")->read() | 0x0100;
+		case 13: return ioport("EFFECT5")->read() | 0x0100;
+		case 14: return ioport("EFFECT6")->read() | 0x0100;
+		case 15: return ioport("EFFECT7")->read() | 0x0100;
```

Comment:
Please use `required_ioport_array` so it doesn’t have to do string hash operations on every access.